src/include/pbs_gss.h

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 1994-2019 Altair Engineering, Inc.
+ * For more information, contact Altair at www.altair.com.
+ *
+ * This file is part of the PBS Professional ("PBS Pro") software.
+ *
+ * Open Source License Information:
+ *
+ * PBS Pro is free software. You can redistribute it and/or modify it under the
+ * terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option) any
+ * later version.
+ *
+ * PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#ifndef PBS_GSS_H
+#define PBS_GSS_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <pbs_config.h>   /* the master config generated by configure */
+
+#if defined(PBS_SECURITY) && (PBS_SECURITY == KRB5)
+
+#include <gssapi.h>
+
+#define gss_nt_service_name GSS_C_NT_HOSTBASED_SERVICE
+#define DIS_GSS_BUF_SIZE 4096 /* default DIS buffer size */
+
+enum PBS_GSS_ROLE {
+	PBS_GSS_ROLE_UNKNOWN = 0,
+	PBS_GSS_CLIENT,
+        PBS_GSS_SERVER,
+        PBS_GSS_ROLE_LAST
+};
+
+typedef struct {
+	gss_ctx_id_t gssctx; /* gss security context */
+	int gssctx_established; /* true if gss context has been established */
+	int ready; /* true if ready to wrap/unwrap message */
+	int confidential; /* wrapping includes encryption */
+	enum PBS_GSS_ROLE role; /* value is client or server */
+	char *hostname; /* server name */
+	char *clientname; /* client name in string */
+	gss_buffer_desc client_name; /* client name in gss buffer */
+	int init_client_ccache; /* if true the client ccache is attempted to be created from keytab */
+	int req_output_size; /* used to determine the wrap_size */
+	OM_uint32 max_input_size; /*  maximum size of an unwrapped message */
+
+	/* TCP only */
+	int establishing; /* true if handshake in progress */
+
+	/* TPP only */
+	char *cleartext; /* saves cleartext for postsend_handler() */
+	int cleartext_len;
+} gss_extra_t;
```

Comment:
The word 'extra' is used for an (extra) structure bonded with TCP or TPP connection. This is the structure that is bonded. I used 'extra' because extra is general and the word appears in the code without GSS (tcparray[fd]->extra or conn->extra) - but only set to NULL. If you have a better idea for the word, I don't mind to change it.